src/dmd/chkprintf.d

```diff
@@ -397,115 +383,34 @@ Format parseFormatSpecifier(scope const char[] format, ref size_t idx,
-    else if (lm == 'h' || lm == 'l')
-    {
-        ++i;
-        if (i == length)
-            return error();
-        lm2 = lm == format[i];
-        if (lm2)
-        {
-            ++i;
-            if (i == length)
-                return error();
-        }
-    }
+    char genSpec;
+    Format specifier = parseGenericFormatSpecifier(format, i, genSpec);
+    if (specifier == Format.error || specifier == Format.lg)
+        return error();
 
-    /* Read the `specifier`
-     */
-    Format specifier;
-    const sc = format[i];
-    ++i;
-    switch (sc)
+    switch (genSpec)
     {
-        case 'd':
-        case 'i':
-        case 'u':
-            if (hash)
-                return error();
-            goto case 'o';
-
-        case 'o':
-        case 'x':
-        case 'X':
-            specifier = lm == 'l' && lm2 ? Format.lld :
-                        lm == 'l'        ? Format.ld  :
-                        lm == 'j'        ? Format.jd  :
-                        lm == 'z'        ? Format.zd  :
-                        lm == 't'        ? Format.td  :
-                                           Format.d;
-            break;
-
-        case 'f':
-        case 'F':
-        case 'e':
-        case 'E':
-        case 'g':
-        case 'G':
-        case 'a':
-        case 'A':
-            if (lm == 'L')
-                specifier = Format.Lg;
-            else if (lm1 ||lm2 || lm == 'h')
-                return error();
-            else
-                specifier = Format.g;
-            break;
-
         case 'c':
-            if (hash || zero ||
-                lm1 || lm2 || lm == 'h')
-                return error();
-            specifier = lm == 'l' ? Format.lc : Format.c;
-            break;
-
         case 's':
-            if (hash || zero ||
-                lm1 || lm2 || lm == 'h')
+            if (hash || zero)
                 return error();
-            specifier = lm == 'l' ? Format.ls : Format.s;
             break;
 
-        case 'p':
-            if (lm1 || lm == 'h' || lm == 'l')
+        case 'd':
+        case 'i':
+            if (hash)
                 return error();
-            specifier = Format.p;
             break;
 
         case 'n':
-            if (flags || hash || zero ||
-                width || precision ||
-                lm == 'L')
-            {
+            if (hash || zero || precision || width || flags)
                 return error();
-            }
-            specifier = lm == 'l' && lm2 ? Format.lln :
-                        lm == 'l'        ? Format.ln  :
-                        lm == 'h' && lm2 ? Format.hhn :
-                        lm == 'h'        ? Format.hn  :
-                        lm == 'j'        ? Format.jn  :
-                        lm == 'z'        ? Format.zn  :
-                        lm == 't'        ? Format.tn  :
-                                           Format.n;
             break;
 
         default:
-            return error();
+            break;
```

Comment:
An undefined specifier should be an error. Please check all "undefined behavior" notations in the C99 spec and make sure they're all covered in the logic and the test cases.

Reply:
Now it's the `parseGenericFormatSpecifier` function (chkformat.d) that checks for undefined behavior.
This `switch` now just checks if there is an incompatible type and specifier (i.e `%#d`). If so, then it returns an error.

Reply:
Now it's the `parseGenericFormatSpecifier` function (chkformat.d) that checks for undefined behavior.
This `switch` now just checks if there is an incompatible type and specifier (i.e `%#d`). If so, then it returns an error.

Reply:
If the `default:` case is unreachable, the code for it should be `assert(0);` not `break;`